### PR TITLE
[cmake] lower minimum cmake version

### PR DESCRIPTION
**Pull request by wgtdkp**
This PR
1. changes the minimum required CMake version from `3.13.1` to `3.10.1` since many platforms (ubuntu, android studio) have 3.10.X as the default CMake version
2. submodule `mdns` is updated to support this change
3. avoid inclusion of unnecessary third parties

### CODE REVIEW

**Comment by jwhui** [comment]
@wgtdkp , should we add a GitHub Action test using CMake 3.10.1?

File: CMakeLists.txt [hunk 1]
```diff
@@ -26,7 +26,7 @@
 #  POSSIBILITY OF SUCH DAMAGE.
 #
 
-cmake_minimum_required(VERSION 3.13.1)
+cmake_minimum_required(VERSION 3.10.1)
```
Comment (jwhui) [inline CMakeLists.txt:29]: @wgtdkp , should we add a GitHub Action test using CMake 3.10.1?

Comment (wgtdkp) [reply CMakeLists.txt:29]: Yes, will do.

**Comment by wgtdkp** [comment]
reopen PR to trigger re-run of GitHub Action.

**Comment by wgtdkp** [comment]
Seems we need to make changes (refactor unittests build) to current CMake files to work with 3.10.1 version. Mark this PR as work-in-progress.

File: src/app/CMakeLists.txt [hunk 2]
```diff
@@ -26,20 +26,17 @@
 #  POSSIBILITY OF SUCH DAMAGE.
 #
 
-add_library(commissioner-app)
-
-target_sources(commissioner-app
```
Comment (wgtdkp) [inline src/app/CMakeLists.txt:31]: CMake 3.10.1 doesn't support `target_sources()`.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -64,23 +64,33 @@ jobs:
...
         run: |
           gcc --version
+          mkdir build && cd build
           cmake -GNinja                           \
                 -DBUILD_SHARED_LIBS=ON            \
                 -DCMAKE_CXX_STANDARD=11           \
                 -DCMAKE_CXX_STANDARD_REQUIRED=ON  \
                 -DCMAKE_BUILD_TYPE=Release        \
                 -DCMAKE_INSTALL_PREFIX=/usr/local \
-                -S . -B build
```
Comment (wgtdkp) [inline .github/workflows/build.yml:79]: CMake 3.10.1 doesn't support `-S` and `-B` options.

File: src/app/CMakeLists.txt [hunk 4]
```diff
@@ -70,25 +67,18 @@ if (BUILD_SHARED_LIBS)
...
+        json_test.cpp
     )
 
     target_include_directories(commissioner-app-test
         PRIVATE
+            ${PROJECT_SOURCE_DIR}/include
             ${PROJECT_SOURCE_DIR}/src
-    )
-
-    target_link_libraries(commissioner-app-test
```
Comment (wgtdkp) [inline src/app/CMakeLists.txt:88]: CMake 3.10.1 doesn't allow `target_link_libraries` on `OBJECT` target.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -64,23 +64,33 @@ jobs:
...
       - uses: actions/checkout@v2
+      - name: Install CMake 3.10.1
+        run: |
+          wget https://cmake.org/files/v3.10/cmake-3.10.1.zip
+          unzip cmake-3.10.1
+          cd cmake-3.10.1
+          cmake .
+          make -j4
+          sudo make install
+          cmake --version
```
Comment (wgtdkp) [inline .github/workflows/build.yml:75]: @jwhui I changed the default gcc+ubuntu build to use CMake `3.10.1`.

**Comment by wgtdkp** [comment]
> Looks good overall. Please help resolve conflicts.

@jwhui done!

**Pull request merged**